### DISCUSSION TITLE
[Help] Firefly + nginx + vouch proxy

### DISCUSSION THREAD

**Discussion by eHonnef** [firefly-iii/firefly-iii#8590] [Firefly III] [answered]
Hello all,

Currently I have a setup where I use NGINX with vouch proxy for authentication (by using auth_request) integrated with keycloak.

Now, I'm trying to run firefly-iii, it has support for [remote user authentication](https://docs.firefly-iii.org/how-to/firefly-iii/advanced/authentication/#remote-user-authentication), but every time the login is successful and it redirects to firefly, I hit the error "The guard header was unexpectedly empty. See the logs.".

Here is a snippet of my relevant NGINX configuration:

```
 auth_request /validate;

 location / {
      proxy_pass_header Authorization;
      proxy_pass http://firefly_node$uri$is_args$args;

       auth_request_set $auth_resp_x_vouch_user $upstream_http_x_vouch_user;
       auth_request_set $auth_resp_x_vouch_idp_claims_groups $upstream_http_x_vouch_idp_claims_groups;
       auth_request_set $auth_resp_x_vouch_idp_claims_given_name $upstream_http_x_vouch_idp_claims_given_name;

        proxy_set_header X-Forwarded-Host $host;
        proxy_set_header X-Forwarded-Server $host;
        proxy_set_header X-Forwarded-Proto $scheme;
        proxy_set_header Host $host;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        client_max_body_size 64M;
        proxy_read_timeout 300s;
        proxy_pass_request_headers on;

        proxy_set_header X-Vouch-User $auth_resp_x_vouch_user;
}    
```

In the `.env`, I've tried both configs: `AUTHENTICATION_GUARD_HEADER=HTTP_REMOTE_USER` and `AUTHENTICATION_GUARD_HEADER=REMOTE_USER`

The vouch service works, it asks for the authentication and forwards to the correct page.

Is there anyone with similar setup that can help? Thanks.

**Comment by JC5** [comment]
In which header will vouch store the authentication details?

**Comment by eHonnef** [comment]
Thanks for the reply,

I’m not very good with this header and proxy logic.
But as I saw in this thread: github.com/vouch/vouch-proxy/issues/421#issuecomment-919520144

That vouch has the user set in “x-vouch-user” from the /validade endpoint, which I van forward to firefly by “proxy_set_header”.

Also, when I try to set the header “remote-user” with the x_vouch_user, my nginx said it is already set.

**Reply by JC5** [reply]
> which I van forward to firefly by “proxy_set_header”.

Could you elaborate a little bit? How do you do this?

x-vouch-user would also work for Firefly III.

**Reply by eHonnef** [reply]
This is my current NGINX config file:

```
location / {
          auth_request_set $auth_resp_x_vouch_user $upstream_http_x_vouch_user;

          proxy_pass http://firefly_node$uri$is_args$args;
          proxy_set_header X-User $auth_resp_x_vouch_user;

          proxy_set_header X-Forwarded-Host $host;
          proxy_set_header X-Forwarded-Server $host;
          proxy_set_header X-Forwarded-Proto $scheme;
          proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;

          client_max_body_size 64M;
          proxy_read_timeout 300s;
          proxy_pass_request_headers off;

          proxy_set_header X-Vouch-User $auth_resp_x_vouch_user;
          proxy_pass_header X-Vouch-User;
}
```

In the NGINX logs I can see that the x-vouch-user was correctly set:

```
<IP_ADDRESS> - - [2024-02-28T19:35:02+01:00] "GET / HTTP/1.0" "<EMAIL_ADDRESS>"
```
If I access `vouch.mydomain.tld/validate`, I will see the header:

![image](https://github.com/firefly-iii/firefly-iii/assets/12504551/6344a4b7-9560-4360-966b-92a3b60b0c8e)



Maybe do I need to change something in the `.env` file?

Currently I have this:

```
AUTHENTICATION_GUARD=remote_user_guard
AUTHENTICATION_GUARD_HEADER=REMOTE_USER
AUTHENTICATION_GUARD_EMAIL=
```

I've tried to set `AUTHENTICATION_GUARD_EMAIL=x-vouch-user`, but no success.

**Reply by JC5** [reply, marked as answer]
https://docs.firefly-iii.org/how-to/firefly-iii/advanced/authentication/#another-remote-user-header

Try `AUTHENTICATION_GUARD_HEADER=HTTP_X_VOUCH_USER`

**Reply by eHonnef** [reply]
It works!!!

Thanks <3
